File: backend/app/services/recommender.py
"""
추천 엔진 서비스 레이어
- tool_recommender.py를 감싸서 API 친화적 결과를 반환
"""
import sys
import os

# 기존 tools 패키지를 import할 수 있도록 경로 추가
_PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
if _PROJECT_ROOT not in sys.path:
    sys.path.insert(0, _PROJECT_ROOT)

from tools.tool_recommender import (
    compute_scores,
    compute_score_breakdown,
    get_persona,
    get_recommendations,
    get_excluded_candidates,
    apply_toxicity_guardrail,
    calculate_bmi,
    get_bmi_category,
    calculate_bmr,
    calculate_tdee,
    calculate_protein_target,
    calculate_monthly_summary,
    apply_family_history_boosts,
    apply_food_detail_boosts,
    apply_drug_interaction_boosts,
    apply_blood_test_boosts,
    SUPPLEMENTS,
    SUPPLEMENT_FOOD_AVOID,
)

from tools.tool_coupang_search import get_coupang_url


def build_coupang_url(keyword: str) -> str:
    """쿠팡 검색 URL 생성 래퍼"""
    result = get_coupang_url(keyword)
    return result.get("url", "")


def run_recommendation(answers: dict) -> dict:
    """
    설문 응답 dict를 받아 전체 추천 파이프라인을 실행하고 결과를 반환.

    이 함수는 supplement_app.py의 step_result()가 하던 일을 하나로 모은 것.
    """
    warnings = []

    # ── 1. 점수 계산 ──
    scores = compute_scores(answers)
    breakdown = compute_score_breakdown(answers)

    # ── 2. 가족력 부스트 ──
    blood_boosts = {}
    extra_supp_boosts = {}

    family_history = answers.get("가족력", [])
    if family_history:
        fam_cat_boosts, fam_supp_boosts, fam_notes = apply_family_history_boosts(family_history)
        for cat, val in fam_cat_boosts.items():
            scores[cat] = scores.get(cat, 0) + val
        extra_supp_boosts.update(fam_supp_boosts)
        warnings.extend(fam_notes)

    # ── 3. 식단 세부 부스트 ──
    food_detail = answers.get("음식상세")
    if food_detail:
        food_boosts, food_warnings = apply_food_detail_boosts(food_detail)
        extra_supp_boosts.update(food_boosts)
        warnings.extend(food_warnings)

    # ── 4. 약물 상호작용 부스트 ──
    drug_boosts, drug_warnings = apply_drug_interaction_boosts(answers)
    extra_supp_boosts.update(drug_boosts)
    warnings.extend(drug_warnings)

    # ── 5. 혈액검사 부스트 ──
    blood_values = answers.get("혈액검사")
    if blood_values:
        blood_boosts = apply_blood_test_boosts(blood_values)

    # ── 6. 페르소나 ──
    persona = get_persona(answers, scores)

    # ── 7. 추천 ──
    current_ids = set(answers.get("현재복용영양제", []))
    raw_recs = get_recommendations(
        answers, scores,
        blood_boosts=blood_boosts,
        max_recs=5,
        extra_supp_boosts=extra_supp_boosts,
        current_ids=current_ids,
    )

    # ── 8. 독성 가드레일 ──
    # raw_recs = [(supp_dict, matched_symptoms, total_score), ...]
    # apply_toxicity_guardrail 은 (filtered_list, warning_messages) 반환
    raw_recs, tox_warnings = apply_toxicity_guardrail(raw_recs, current_ids=current_ids, max_recs=5)
    warnings.extend(tox_warnings)

    # ── 9. 제외 후보 ──
    # get_excluded_candidates 는 원본 튜플 형식을 기대
    raw_excluded = get_excluded_candidates(
        answers, scores, raw_recs,
        blood_boosts=blood_boosts,
        extra_supp_boosts=extra_supp_boosts,
        top_n=3,
    )

    # ── 10. 영양 정보 ──
    height = answers.get("신장", 170)
    weight = answers.get("체중", 65)
    gender = answers.get("성별", "male")
    age = answers.get("나이", 30)
    activity = answers.get("운동", "거의_안함")

    bmi = calculate_bmi(weight, height)
    bmi_label, bmi_color, bmi_advice = get_bmi_category(bmi)
    bmr = calculate_bmr(gender, weight, height, age)
    tdee = calculate_tdee(bmr, activity)
    protein_min, protein_max = calculate_protein_target(weight, answers.get("목표", []))

    # ── 11. 튜플 → API dict 변환 ──
    # raw_recs: [(supp_dict, matched_symptoms, total_score), ...]
    recs = []
    for rank, (supp, matched_syms, total_score) in enumerate(raw_recs, 1):
        supp_id = supp.get("id", "")
        recs.append({
            "id": supp_id,
            "name": supp.get("name", supp_id),
            "score": total_score,
            "rank": rank,
            "mfds_type": supp.get("mfds_type"),
            "mfds_function": supp.get("mfds_function"),
            "evidence": supp.get("evidence"),
            "dosage_guide": supp.get("dosage_guide"),
            "cautions": supp.get("cautions", []),
            "drug_interactions": supp.get("drug_interactions", []),
            "symptom_indicators": supp.get("symptom_indicators", []),
            "coupang_url": build_coupang_url(supp.get("name", supp_id)),
            "food_avoid": SUPPLEMENT_FOOD_AVOID.get(supp_id, []),
        })

    # raw_excluded: [(supp_dict, reason_str), ...]
    excluded = []
    for supp, reason in raw_excluded:
        supp_id = supp.get("id", "")
        excluded.append({
            "id": supp_id,
            "name": supp.get("name", supp_id),
            "score": 0,
            "rank": 0,
            "reason": reason,
            "coupang_url": build_coupang_url(supp.get("name", supp_id)),
            "food_avoid": SUPPLEMENT_FOOD_AVOID.get(supp_id, []),
        })

    # ── 12. 월간 비용 (API dict 형식으로 전달) ──
    monthly = calculate_monthly_summary(raw_recs)

    # ── 결과 조립 ──
    return {
        "persona": persona,
        "scores": scores,
        "score_breakdown": [
            {"category": cat, "score": sc, "max_score": 10.0}
            for cat, sc in breakdown.items()
        ],
        "recommendations": recs,
        "excluded": excluded,
        "nutrition_info": {
            "bmi": {
                "value": round(bmi, 1),
                "label": bmi_label,
                "color": bmi_color,
                "advice": bmi_advice,
            },
            "bmr": round(bmr, 0),
            "tdee": round(tdee, 0),
            "protein_min": round(protein_min, 0),
            "protein_max": round(protein_max, 0),
        },
        "monthly_summary": monthly,
        "warnings": warnings,
    }

File: web/src/engine/recommender.ts
/**
 * 영양제 추천 엔진
 */

import type { SurveyAnswers, RecommendationResult, Supplement, Persona, NutritionInfo, MonthlySummary } from '../types';
import type { SupplementData, PersonaData } from './data';
import {
  SUPPLEMENTS,
  PERSONAS,
  SUPPLEMENT_MONTHLY_DATA,
  SUPPLEMENT_FOOD_AVOID,
  COUPANG_KEYWORDS,
  COUPANG_PARTNER_LINKS,
} from './data';
import {
  computeScores,
  calculateBMI,
  getBMICategory,
  calculateBMR,
  calculateTDEE,
  calculateProteinTarget,
} from './scorer';

function toPersona(p: PersonaData): Persona {
  return {
    id: p.id,
    name: p.name,
    emoji: p.emoji,
    description: p.description || p.tagline || '',
    tagline: p.tagline || '',
    tip: p.tip || '',
  };
}

export function getPersona(answers: SurveyAnswers, scores: Record<string, number>): Persona {
  const findPersona = (id: string): Persona => {
    const found = PERSONAS.find((p) => p.id === id);
    return toPersona(found || PERSONAS[PERSONAS.length - 1]);
  };

  const age = answers.나이 || 30;
  const gender = answers.성별 || 'male';
  const conditions = answers.기저질환 || [];
  const alcohol = answers.음주 || '거의_안함';
  const exercise = answers.운동 || '거의_안함';

  // 절대 우선순위
  if (conditions.includes('임산부_수유부')) return findPersona('pregnant');

  // 카테고리 기반 매칭
  const sorted = Object.entries(scores).sort((a, b) => b[1] - a[1]);
  const topCat = sorted[0]?.[0];
  const totalScore = Object.values(scores).reduce((s, v) => s + v, 0);

  // 증상이 1개 이하이면 전반적 건강 관리형
  const symptomCount = (answers.증상 || []).length;
  if (totalScore < 8 || symptomCount <= 1) return findPersona('general_wellness');

  const topScore = sorted[0]?.[1] || 0;
  const scoreOf = (cat: string) => scores[cat] || 0;
  const isFemale = gender === 'female' || gender === '여성';
  const isMale = gender === 'male' || gender === '남성';
  const drinksOften = alcohol === '주3-4회' || alcohol === '매일';
  const exercisesOften = exercise === '주3-4회' || exercise === '거의_매일';

  // ── 우선 체크: 음주자 + 간건강 점수가 상위권이면 즉흥 외식형 ──
  if (drinksOften && scoreOf('간건강') >= topScore - 2) return findPersona('dine_out_drinker');
  if (alcohol !== '거의_안함' && topCat === '간건강') return findPersona('dine_out_drinker');

  // ── 우선 체크: 운동 자주 + 근육관절 상위권이면 근력 마니아 ──
  if (exercisesOften && scoreOf('근육관절') >= topScore - 2) return findPersona('fitness_lover');

  // 간건강 — 비음주자 (불규칙 생활)
  if (topCat === '간건강') return findPersona('low_willpower');

  // 갱년기
  if (topCat === '갱년기' && isFemale) return findPersona('menopause_woman');
  if (topCat === '갱년기' && isMale && age >= 40) return findPersona('middle_aged_man');

  // 혈당대사
  if (topCat === '혈당대사') return findPersona('blood_sugar_manager');

  // 심혈관
  if (topCat === '심혈관' && isMale && age >= 40) return findPersona('middle_aged_man');
  if (topCat === '심혈관') return findPersona(age >= 50 ? 'senior_wellness' : 'burnout');

  // 수면
  if (topCat === '수면') return findPersona('poor_sleeper');

  // 스트레스 / 피로
  if (topCat === '스트레스') return findPersona('burnout');
  if (topCat === '피로') return findPersona('burnout');

  // 근육관절 (운동 안 하는 경우)
  if (topCat === '근육관절') return findPersona(age >= 50 ? 'senior_wellness' : 'burnout');

  // 피부
  if (topCat === '피부') return findPersona('skin_influencer');

  // 장건강
  if (topCat === '장건강') return findPersona('gut_health_sensitive');

  // 면역력
  if (topCat === '면역력') return findPersona('immunity_worrier');

  // 눈건강
  if (topCat === '눈건강') return findPersona('burnout');

  // 체중관리
  if (topCat === '체중관리') return findPersona('dieter');

  // 인지기능
  if (topCat === '인지기능') return findPersona(age >= 50 ? 'senior_wellness' : 'burnout');

  // 나이 기반 폴백
  if (age >= 50) return findPersona('senior_wellness');

  return findPersona('general_wellness');
}

export function getRecommendations(
  answers: SurveyAnswers,
  scores: Record<string, number>,
  persona: PersonaData | null,
  options?: { current_ids?: Set<string>; max_recs?: number },
): Array<[SupplementData, string[]]> {
  const { current_ids = new Set(), max_recs = persona?.max_recommendations || 5 } = options || {};
  const symptoms = answers.증상 || [];
  const conditions = answers.기저질환 || [];

  const scored: Array<{ total: number; supp: SupplementData; matched: string[] }> = [];

  for (const supp of SUPPLEMENTS) {
    // 차단 조건 체크
    if (current_ids.has(supp.id)) continue;
    if (supp.blocked_conditions?.some((c: string) => conditions.includes(c))) continue;

    // 카테고리 점수 × 친화도
    let total = 0;
    for (const [cat, weight] of Object.entries(supp.affinity || {})) {
      total += (scores[cat] || 0) * (weight as number);
    }

    // 성별 부스트
    if (supp.gender_boost) {
      total += supp.gender_boost[answers.성별] || 0;
    }

    // 음주 부스트
    if (supp.drink_boost && answers.음주 && answers.음주 !== '거의_안함') {
      total += supp.drink_boost;
    }

    // 증상 매칭 보너스
    const matched: string[] = [];
    if (supp.symptom_indicators) {
      for (const ind of supp.symptom_indicators) {
        if (symptoms.includes(ind)) {
          total += 8;
          matched.push(ind);
        }
      }
    }

    if (total >= 12) {
      scored.push({ total, supp, matched });
    }
  }

  scored.sort((a, b) => b.total - a.total);

  // 시너지 보너스
  const topIds = new Set(scored.slice(0, max_recs + 2).map((s) => s.supp.id));
  for (const item of scored) {
    if (item.supp.synergy) {
      for (const syn of item.supp.synergy) {
        if (topIds.has(syn.id)) {
          item.total += 5;
        }
      }
    }
  }

  scored.sort((a, b) => b.total - a.total);

  return scored.slice(0, max_recs).map((s) => [s.supp, s.matched]);
}

export function calculateMonthlySummary(recommendations: Array<[SupplementData, string[]]>): MonthlySummary {
  let cost_min = 0;
  let cost_max = 0;
  let pills_per_day = 0;
  const items: Array<{ name: string; pills: number; unit: string }> = [];

  for (const [supp] of recommendations) {
    const data = SUPPLEMENT_MONTHLY_DATA[supp.id];
    if (data) {
      cost_min += data.cost_min;
      cost_max += data.cost_max;
      pills_per_day += data.pills_per_day;
      items.push({ name: supp.name, pills: data.pills_per_day, unit: data.unit });
    }
  }

  return { cost_min, cost_max, pills_per_day, items };
}

export function runRecommendation(answers: SurveyAnswers): RecommendationResult {
  const scores = computeScores(answers);
  const persona = getPersona(answers, scores);

  const personaData = PERSONAS.find((p) => p.id === persona.id) || null;
  const current_ids = new Set(answers.현재복용영양제 || []);
  const raw_recs = getRecommendations(answers, scores, personaData, { current_ids });

  // 신체 정보
  const height = answers.신장 || 170;
  const weight = answers.체중 || 65;
  const gender = answers.성별 || 'male';
  const age = answers.나이 || 30;
  const activity = answers.운동 || '거의_안함';

  const bmi = calculateBMI(weight, height);
  const { label: bmi_label, color: bmi_color, advice: bmi_advice } = getBMICategory(bmi);
  const bmr = calculateBMR(gender, weight, height, age);
  const tdee = calculateTDEE(bmr, activity);
  const [protein_min, protein_max] = calculateProteinTarget(weight, answers.목표 || []);

  const nutrition_info: NutritionInfo = {
    bmi: { value: Math.round(bmi * 10) / 10, label: bmi_label, color: bmi_color, advice: bmi_advice },
    bmr: Math.round(bmr),
    tdee: Math.round(tdee),
    protein_min: Math.round(protein_min),
    protein_max: Math.round(protein_max),
  };

  const monthly_summary = calculateMonthlySummary(raw_recs);

  const recommendations: Supplement[] = raw_recs.map(([supp, matched], idx) => ({
    id: supp.id,
    name: supp.name,
    score: 0,
    rank: idx + 1,
    mfds_type: supp.evidence?.mfds_type || supp.evidence?.summary || '',
    mfds_function: supp.evidence?.mfds_function || '',
    evidence: supp.evidence ? {
      summary: supp.evidence.summary || supp.evidence.mfds_function || '',
      studies: supp.evidence.study ? [supp.evidence.study] : [],
    } : undefined,
    dosage_guide: supp.dosage_guide ? {
      amount: supp.dosage_guide.amount || '',
      timing: supp.dosage_guide.timing || '',
      duration: supp.onset_weeks || '',
    } : undefined,
    cautions: supp.cautions || [],
    drug_interactions: supp.drug_interactions || [],
    symptom_indicators: matched,
    coupang_url: COUPANG_PARTNER_LINKS[supp.id] || `https://www.coupang.com/np/search?q=${encodeURIComponent(COUPANG_KEYWORDS[supp.id] || supp.name + ' 건강기능식품')}&channel=user&sorter=scoreDesc`,
    food_avoid: (SUPPLEMENT_FOOD_AVOID[supp.id] || []).map((f: any) => typeof f === 'string' ? f : f.item || ''),
  }));

  return {
    persona,
    scores,
    score_breakdown: Object.entries(scores).map(([category, score]) => ({
      category,
      score,
      max_score: 10.0,
    })),
    recommendations,
    excluded: [],
    nutrition_info,
    monthly_summary,
    warnings: [],
  };
}
